inference-engine/src/mkldnn_plugin/nodes/mkldnn_pad_node.cpp

```diff
@@ -0,0 +1,336 @@
+    float* dstData = getChildEdgeAt(0)->getBlob()->buffer().as<float*>() +
+            getChildEdgeAt(0)->getBlob()->getTensorDesc().getBlockingDesc().getOffsetPadding();
+
+    auto srcDims = getParentEdgeAt(0)->getDims().ToSizeVector();
+    auto dstDims = getChildEdgeAt(0)->getDims().ToSizeVector();
+
+    switch (padMode) {
+        case CONSTANT:
+            padConstant(srcData, dstData, srcDims, dstDims);
+            break;
+        case EDGE:
+            padEdge(srcData, dstData, srcDims, dstDims);
+            break;
+        case REFLECT:
+            padReflect(srcData, dstData, srcDims, dstDims);
+            break;
+        case SYMMETRIC:
+            padSymmetric(srcData, dstData, srcDims, dstDims);
+            break;
+    }
+}
+
+inline size_t parallel_init(size_t start, size_t size, std::vector<size_t> &counters, std::vector<size_t> &dims) {
+    for (int j = size - 1; j >= 0; j--) {
+        counters[j] = start % dims[j];
+        start = start / dims[j];
+    }
+    return start;
+}
+
+inline void parallel_step(size_t size, std::vector<size_t> &counters, std::vector<size_t> &dims) {
+    for (int j = size - 1; j >= 0; j--) {
+        counters[j] = (counters[j] + 1) % dims[j];
+        if (counters[j] != 0)
+            return;
+    }
+}
+
+void MKLDNNPadNode::padConstant(const float* srcData, float* dstData, SizeVector srcDims, SizeVector dstDims) {
+    size_t dimsSize_1 = dstDims.size() - 1;
+    size_t inputSV = srcDims[dimsSize_1];
+    size_t workAmountSrc = params.srcStrides[0] * srcDims[0] / srcDims[dimsSize_1];
+
+    int offset = 0;
+    for (size_t i = 0; i < params.srcStrides.size(); ++i)
+        offset += padsBegin[i] * params.dstStrides[i];
+    std::fill_n(dstData, offset, padValue);
+
+    parallel_nt(0, [&](const int ithr, const int nthr) {
+        size_t start = 0, end = 0;
+        SizeVector counters(dimsSize_1, 0);
+        splitter(workAmountSrc, nthr, ithr, start, end);
+        SizeVector countersNext(dimsSize_1, 0);
+
+        parallel_init(start, dimsSize_1, counters, srcDims);
+        parallel_init(start, dimsSize_1, countersNext, srcDims);
+        parallel_step(dimsSize_1, countersNext, srcDims);
+        int srcIdx = 0;
+        int dstIdx = padsBegin[dimsSize_1];
+        int dstIdxNext = padsBegin[dimsSize_1];
+        for (size_t i = 0; i < dimsSize_1; ++i) {
+            srcIdx += counters[i] * params.srcStrides[i];
+            dstIdx += (padsBegin[i] + counters[i]) * params.dstStrides[i];
+            dstIdxNext += (padsBegin[i] + countersNext[i]) * params.dstStrides[i];
+        }
+        if (dstIdxNext <= dstIdx) dstIdxNext = params.dstStrides[0] * dstDims[0];
+
+        for (size_t iwork = start; iwork < end; ++iwork, srcIdx += inputSV) {
+            cpu_memcpy(&dstData[dstIdx], &srcData[srcIdx], sizeof(float) * inputSV);
+            std::fill_n(&dstData[dstIdx + inputSV], dstIdxNext - dstIdx - inputSV, padValue);
```

Comment:
If I understand correctly this approach assumes that two consecutive rows should lay without any stride between them into memory. In general this is not very good approach from dynamic shapes support standpoint. `padEdge` version looks more generic and do almost the same stuff, but with different value to fill padding. 
Can we check that `padEdge` shows performance not worse than `padConstant`? In that case both implementations can be unite. 
cc'ed @alexey-varyzgin